src/DistributedComputations/distributed_architectures.jl

```diff
@@ -6,55 +6,109 @@ import Oceananigans.Architectures: device, cpu_architecture, arch_array, array_t
 import Oceananigans.Grids: zeros
 import Oceananigans.Utils: sync_device!, tupleit
 
+import Base
+
 #####
 ##### Partitioning
 #####
 
-# Form 3-tuple
-regularize_size(sz::Tuple{<:Int, <:Int}, Rx, Ry) = (sz[1], sz[2], 1)
-
-# Infer whether x- or y-dimension is indicated
-function regularize_size(N::Int, Rx, Ry)
-    if Rx == 1
-        return (1, N, 1)
-    elseif Ry == 1
-        return (N, 1, 1)
-    else
-        throw(ArgumentError("We can't interpret the 1D size $N for 2D partitions!"))
-    end
+struct Partition{Sx, Sy, Sz}
+    x :: Sx
+    y :: Sy
+    z :: Sz
 end
 
-regularize_size(sz::Array{<:Int, 1}) = regularize_size(sz[1])
-regularize_size(sz::Array{<:Int, 2}) = regularize_size((sz[1], sz[2]))
-regularize_size(sz::Tuple{<:Int})    = regularize_size(sz[1])
+"""
+    Partition(; x = 1, y = 1, z = 1)
+
+Return `Partition` representing the division of a domain in
+the `x` (first), `y` (second) and `z` (third) dimension
+
+Keyword arguments: 
+==================
+
+- `x`: partitioning of the first dimension 
+- `y`: partitioning of the second dimension
+- `z`: partitioning of the third dimension
+
+if supplied as positional arguments `x` will be the first argument, 
+`y` the second and `z` the third
+
+`x`, `y` and `z` can be `Int`, `Equal`, `Fractional` or `Sizes` 
+(see below)
+
+Examples:
+========
+
+```jldoctest
+julia> using Oceananigans; using Oceananigans.DistributedComputations
+
+julia> Partition(1, 4)
+Domain partitioning with (1, 4, 1) ranks
+├── x-partitioning: none
+├── y-partitioning: 4
+└── z-partitioning: none
+
+julia> Partition(x = Fractional(1, 2, 3, 4))
+Domain partitioning with (4, 1, 1) ranks
+├── x-partitioning: domain fractions: (0.1, 0.2, 0.3, 0.4)
+├── y-partitioning: none
+└── z-partitioning: none
+
+```
+"""
+Partition(x)       = Partition(validate_partition(x, 1, 1)...)
+Partition(x, y)    = Partition(validate_partition(x, y, 1)...)
+
+Partition(; x = 1, y = 1, z = 1) = Partition(validate_partition(x, y, z)...)
+
+Base.show(io::IO, p::Partition) =
+    print(io,
+    "Domain partitioning with $(ranks(p)) ranks", "\n",
+    "├── x-partitioning: $(ranks(p.x) == 1 ? "none" : p.x)", "\n",
+    "├── y-partitioning: $(ranks(p.y) == 1 ? "none" : p.y)", "\n",
+    "└── z-partitioning: $(ranks(p.z) == 1 ? "none" : p.z)")
 
-# For 3D partitions:
-# regularize_size(sz::Tuple{<:Int, <:Int, <:Int}, Rx, Ry) = sz
+"""type representing equal domain partitioning (not supported for more than one direction)"""
+struct Equal end
```

Comment:
I think we need a real docstring because this is user facing, something like

```suggestion
"""
    Equal()

Return a type that partitions a direction equally among remaining processes.

`Equal()` can be used for only one direction. Other directions must either be unspecified, or
specifically defined by `Int`, `Fractional`, or `Sizes`.
"""
```